File: den_memories/scoring.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from .registry import load_scoring_defaults

_DEFAULTS = load_scoring_defaults()

AUTHORITATIVE_MATCH_WEIGHT = float(_DEFAULTS["root_match"]["authoritative_scope_match"])
DISCOVERABLE_MATCH_WEIGHT = float(_DEFAULTS["root_match"]["discoverable_scope_match"])
EXPLICIT_ONLY_MATCH_WEIGHT = float(_DEFAULTS["root_match"]["explicit_only_match"])
SAME_PROJECT_MATCH_WEIGHT = float(_DEFAULTS["root_match"]["same_project_scope_match"])
LINKED_PROJECT_MATCH_WEIGHT = float(_DEFAULTS["root_match"]["linked_project_scope_match"])
CROSS_SCOPE_PENALTY = float(_DEFAULTS["penalties"]["cross_scope_discovered_evidence"])
AUTHORITY_SCOPE_MISMATCH_PENALTY = float(_DEFAULTS["penalties"]["authority_scope_mismatch"])
CLAIM_STRENGTH_POLICY_WEIGHT = float(_DEFAULTS["claim_strength_modifier"]["policy"])
CLAIM_STRENGTH_RECOMMENDATION_WEIGHT = float(_DEFAULTS["claim_strength_modifier"]["recommendation"])
CLAIM_STRENGTH_ASSESSMENT_WEIGHT = float(_DEFAULTS["claim_strength_modifier"]["assessment"])
CLAIM_STRENGTH_OBSERVATION_WEIGHT = float(_DEFAULTS["claim_strength_modifier"]["observation"])
CURATED_STATUS_WEIGHT = float(_DEFAULTS["curation_modifier"]["curated"])
SOURCE_VERIFIED_WEIGHT = float(_DEFAULTS["source_validation_modifier"]["verified"])
SOURCE_BROKEN_PENALTY = float(_DEFAULTS["source_validation_modifier"]["broken"])
EDGE_TRAVERSAL_DECAY = float(_DEFAULTS["traversal"]["edge_traversal_decay"])
SUPERSEDED_STATUS_PENALTY = float(_DEFAULTS["penalties"]["superseded_status"])
DEPRECATED_STATUS_PENALTY = float(_DEFAULTS["penalties"]["deprecated_status"])
ARCHIVED_STATUS_PENALTY = float(_DEFAULTS["penalties"]["archived_status"])

CLAIM_STRENGTH_WEIGHTS = {
    "observation": CLAIM_STRENGTH_OBSERVATION_WEIGHT,
    "assessment": CLAIM_STRENGTH_ASSESSMENT_WEIGHT,
    "recommendation": CLAIM_STRENGTH_RECOMMENDATION_WEIGHT,
    "policy": CLAIM_STRENGTH_POLICY_WEIGHT,
}

STATUS_WEIGHTS = {key: float(value) for key, value in _DEFAULTS["entry_status_modifier"].items()}

CONFIDENCE_WEIGHTS = {key: float(value) for key, value in _DEFAULTS["confidence_modifier"].items()}
SOURCE_VALIDATION_WEIGHTS = {key: float(value) for key, value in _DEFAULTS["source_validation_modifier"].items()}
EDGE_RELATION_WEIGHTS = {key: float(value) for key, value in _DEFAULTS["edge_relation_modifier"].items()}

EXCLUDED_ENTRY_STATUSES = {"superseded", "deprecated", "archived"}


@dataclass(frozen=True)
class RecallContext:
    scope_kind: str = "global"
    scope_id: str | None = None


def authority_matches(item: dict[str, Any], context: RecallContext) -> bool:
    return item.get("authority_scope_kind") == context.scope_kind and item.get("authority_scope_id") == context.scope_id


def is_global_authority(item: dict[str, Any]) -> bool:
    return item.get("authority_scope_kind") == "global" and not item.get("authority_scope_id")


def is_discoverable(item: dict[str, Any], context: RecallContext) -> bool:
    discovery = item.get("discovery_scope", "explicit_only")
    if authority_matches(item, context) or is_global_authority(item):
        return True
    if discovery == "global_discoverable":
        return True
    if discovery == "same_project" and context.scope_kind == item.get("scope_kind") == "project" and context.scope_id == item.get("scope_id"):
        return True
    if discovery == "linked_projects" and context.scope_kind == "project":
        return True
    return False


def authority_label(item: dict[str, Any], context: RecallContext) -> str:
    if authority_matches(item, context) or is_global_authority(item):
        return "authoritative"
    if is_discoverable(item, context):
        return "discovered_evidence"
    return "out_of_scope"


def source_validation_score(source_statuses: list[str]) -> float:
    if not source_statuses:
        return SOURCE_VALIDATION_WEIGHTS.get("unverified", 0.0)
    return sum(SOURCE_VALIDATION_WEIGHTS.get(status, 0.0) for status in source_statuses) / len(source_statuses)


def score_item(item: dict[str, Any], context: RecallContext, source_statuses: list[str] | None = None, *, edge_depth: int = 0, relation: str | None = None) -> dict[str, Any]:
    label = authority_label(item, context)
    if label == "authoritative":
        score = AUTHORITATIVE_MATCH_WEIGHT
    elif label == "discovered_evidence":
        score = DISCOVERABLE_MATCH_WEIGHT + CROSS_SCOPE_PENALTY
    else:
        score = AUTHORITY_SCOPE_MISMATCH_PENALTY
    score += CLAIM_STRENGTH_WEIGHTS.get(item.get("claim_strength", "observation"), 0.0)
    score += STATUS_WEIGHTS.get(item.get("status", "active"), 0.0)
    score += CONFIDENCE_WEIGHTS.get(item.get("confidence", "unverified"), 0.0)
    score += source_validation_score(source_statuses or [])
    if relation:
        score += EDGE_RELATION_WEIGHTS.get(relation, 0.0)
    if edge_depth:
        score *= EDGE_TRAVERSAL_DECAY ** edge_depth
    return {"score": round(score, 3), "authority_label": label}


def scoring_defaults_readback() -> dict[str, Any]:
    return {
        "profile": _DEFAULTS["profile"],
        "description": _DEFAULTS["description"],
        "AUTHORITATIVE_MATCH_WEIGHT": AUTHORITATIVE_MATCH_WEIGHT,
        "DISCOVERABLE_MATCH_WEIGHT": DISCOVERABLE_MATCH_WEIGHT,
        "CROSS_SCOPE_PENALTY": CROSS_SCOPE_PENALTY,
        "CLAIM_STRENGTH_POLICY_WEIGHT": CLAIM_STRENGTH_POLICY_WEIGHT,
        "CLAIM_STRENGTH_RECOMMENDATION_WEIGHT": CLAIM_STRENGTH_RECOMMENDATION_WEIGHT,
        "CLAIM_STRENGTH_ASSESSMENT_WEIGHT": CLAIM_STRENGTH_ASSESSMENT_WEIGHT,
        "CLAIM_STRENGTH_OBSERVATION_WEIGHT": CLAIM_STRENGTH_OBSERVATION_WEIGHT,
        "CURATED_STATUS_WEIGHT": CURATED_STATUS_WEIGHT,
        "SOURCE_VERIFIED_WEIGHT": SOURCE_VERIFIED_WEIGHT,
        "SOURCE_BROKEN_PENALTY": SOURCE_BROKEN_PENALTY,
        "EDGE_TRAVERSAL_DECAY": EDGE_TRAVERSAL_DECAY,
    }

File: internal/recall/scoring.go
// Package recall implements guided recall scoring and packet helpers.
package recall

import (
	"fmt"
	"math"

	"den-memories/internal/contracts"
	"den-memories/internal/store"
)

// Context describes the authority scope for a recall request.
type Context struct {
	ScopeKind string
	ScopeID   string
}

// Scoring contains v0 recall scoring constants.
type Scoring struct {
	Defaults map[string]any
	Weights  map[string]float64
}

// LoadScoring loads scoring defaults from contract artifacts.
func LoadScoring(loader *contracts.Loader) (*Scoring, error) {
	defaults, err := loader.ScoringDefaults()
	if err != nil {
		return nil, err
	}
	s := &Scoring{Defaults: defaults, Weights: map[string]float64{}}
	for section, values := range defaults {
		items, ok := values.(map[string]any)
		if !ok {
			continue
		}
		for key, value := range items {
			if f, ok := value.(float64); ok {
				s.Weights[section+"."+key] = f
			}
		}
	}
	return s, nil
}

// ScoreItem computes the v0 recall score and authority label for an item.
func (s *Scoring) ScoreItem(item store.Record, ctx Context, sourceStatuses []string, edgeDepth int, relation string) map[string]any {
	label := authorityLabel(item, ctx)
	score := 0.0
	switch label {
	case "authoritative":
		score = s.weight("root_match.authoritative_scope_match")
	case "discovered_evidence":
		score = s.weight("root_match.discoverable_scope_match") + s.weight("penalties.cross_scope_discovered_evidence")
	default:
		score = s.weight("penalties.authority_scope_mismatch")
	}
	score += s.weight("claim_strength_modifier." + stringValue(item["claim_strength"], "observation"))
	score += s.weight("entry_status_modifier." + stringValue(item["status"], "active"))
	score += s.weight("confidence_modifier." + stringValue(item["confidence"], "unverified"))
	score += s.sourceValidationScore(sourceStatuses)
	if relation != "" {
		score += s.weight("edge_relation_modifier." + relation)
	}
	if edgeDepth > 0 {
		score *= math.Pow(s.weight("traversal.edge_traversal_decay"), float64(edgeDepth))
	}
	return map[string]any{"score": math.Round(score*1000) / 1000, "authority_label": label}
}

// Readback returns the public scoring default readback shape.
func (s *Scoring) Readback() map[string]any {
	return map[string]any{
		"profile":                              s.Defaults["profile"],
		"description":                          s.Defaults["description"],
		"AUTHORITATIVE_MATCH_WEIGHT":           s.weight("root_match.authoritative_scope_match"),
		"DISCOVERABLE_MATCH_WEIGHT":            s.weight("root_match.discoverable_scope_match"),
		"CROSS_SCOPE_PENALTY":                  s.weight("penalties.cross_scope_discovered_evidence"),
		"CLAIM_STRENGTH_POLICY_WEIGHT":         s.weight("claim_strength_modifier.policy"),
		"CLAIM_STRENGTH_RECOMMENDATION_WEIGHT": s.weight("claim_strength_modifier.recommendation"),
		"CLAIM_STRENGTH_ASSESSMENT_WEIGHT":     s.weight("claim_strength_modifier.assessment"),
		"CLAIM_STRENGTH_OBSERVATION_WEIGHT":    s.weight("claim_strength_modifier.observation"),
		"CURATED_STATUS_WEIGHT":                s.weight("curation_modifier.curated"),
		"SOURCE_VERIFIED_WEIGHT":               s.weight("source_validation_modifier.verified"),
		"SOURCE_BROKEN_PENALTY":                s.weight("source_validation_modifier.broken"),
		"EDGE_TRAVERSAL_DECAY":                 s.weight("traversal.edge_traversal_decay"),
	}
}

func (s *Scoring) sourceValidationScore(statuses []string) float64 {
	if len(statuses) == 0 {
		return s.weight("source_validation_modifier.unverified")
	}
	total := 0.0
	for _, status := range statuses {
		total += s.weight("source_validation_modifier." + status)
	}
	return total / float64(len(statuses))
}

func (s *Scoring) weight(key string) float64 {
	value, ok := s.Weights[key]
	if !ok {
		return 0
	}
	return value
}

func authorityLabel(item store.Record, ctx Context) string {
	if authorityMatches(item, ctx) || isGlobalAuthority(item) {
		return "authoritative"
	}
	if isDiscoverable(item, ctx) {
		return "discovered_evidence"
	}
	return "out_of_scope"
}

func authorityMatches(item store.Record, ctx Context) bool {
	return stringValue(item["authority_scope_kind"], "") == ctx.ScopeKind && stringValue(item["authority_scope_id"], "") == ctx.ScopeID
}

func isGlobalAuthority(item store.Record) bool {
	return stringValue(item["authority_scope_kind"], "global") == "global" && stringValue(item["authority_scope_id"], "") == ""
}

func isDiscoverable(item store.Record, ctx Context) bool {
	if authorityMatches(item, ctx) || isGlobalAuthority(item) {
		return true
	}
	discovery := stringValue(item["discovery_scope"], "explicit_only")
	if discovery == "global_discoverable" {
		return true
	}
	if discovery == "same_project" && ctx.ScopeKind == "project" && stringValue(item["scope_kind"], "") == "project" && stringValue(item["scope_id"], "") == ctx.ScopeID {
		return true
	}
	return discovery == "linked_projects" && ctx.ScopeKind == "project"
}

func stringValue(value any, fallback string) string {
	if value == nil {
		return fallback
	}
	if text, ok := value.(string); ok {
		return text
	}
	return fmt.Sprint(value)
}
